### PR TITLE
Mock list all

### PR DESCRIPTION
**Pull request by saroad2**
An implementation of the function `listAll` for `MockReference` 

Related to #32

### CODE REVIEW

**Comment by saroad2** [comment]
Hey @atn832 , would love your feedback on this!
My application tests are failing due to the lack of this functionality 🥺

File: test/firebase_storage_mocks_test.dart [hunk 1]
```diff
@@ -115,9 +118,46 @@ void main() {
...
     });
+
+    test('Ref listAll', () async {
+      final basePath = 'this/is/basic';
+      final otherPath = 'another/path';
+      final storage = MockFirebaseStorage();
+      await storage.ref(basePath + '/data2').putData(randomData(255));
+      await storage.ref(basePath + '/subdir1/data1').putData(randomData(255));
+      await storage.ref(basePath + '/data2').putData(randomData(255));
+      await storage.ref(otherPath + '/data3').putData(randomData(255));
```
Comment (atn832) [inline test/firebase_storage_mocks_test.dart:129]: It'd have been nice to use vary the file names a little to make the tests easier to read. We do expect a file called data3, but it's the one from basePath.

File: test/firebase_storage_mocks_test.dart [hunk 2]
```diff
@@ -115,9 +118,46 @@ void main() {
...
+      expectRef(listResult.prefixes[1], name: 'subdir2');
+      expect(listResult.items.length, 3);
+      expectRef(listResult.items[0], name: 'data2');
+      expectRef(listResult.items[1], name: 'data3');
+      expectRef(listResult.items[2], name: 'file3');
+    });
   });
 }
 
+void expectRef(Reference actualReference, {required String name}) {
```
Comment (atn832) [inline test/firebase_storage_mocks_test.dart:151]: Nice!

File: test/firebase_storage_mocks_test.dart [hunk 3]
```diff
@@ -115,9 +118,46 @@ void main() {
...
+  expect(actualReference,
+      isA<MockReference>().having((ref) => ref.name, 'Right name', name));
+}
+
+Uint8List randomData(int n) {
```
Comment (atn832) [inline test/firebase_storage_mocks_test.dart:156]: Nice!

File: lib/src/mock_storage_reference.dart [hunk 4]
```diff
@@ -89,6 +90,33 @@ class MockReference implements Reference {
...
+    final prefixes = <String>[], items = <String>[];
+    final allPaths = <String>[
+      ..._storage.storedDataMap.keys,
+      ..._storage.storedFilesMap.keys
+    ];
+    for (var child in allPaths) {
+      if (!child.startsWith(normalizedPath)) continue;
+      final relativeChild = child.substring(normalizedPath.length);
+      if (relativeChild.contains('/')) {
+        var prefix = normalizedPath + relativeChild.split('/')[0];
```
Comment (atn832) [inline lib/src/mock_storage_reference.dart:105]: Seems like this could be final.

File: lib/src/mock_storage_reference.dart [hunk 5]
```diff
@@ -89,6 +90,33 @@ class MockReference implements Reference {
     return Future.value(_storage.storedDataMap[_path]);
   }
 
+  @override
+  Future<ListResult> listAll() {
+    final normalizedPath = _path.endsWith('/') ? _path : _path + '/';
+    final prefixes = <String>[], items = <String>[];
+    final allPaths = <String>[
+      ..._storage.storedDataMap.keys,
+      ..._storage.storedFilesMap.keys
```
Comment (atn832) [inline lib/src/mock_storage_reference.dart:99]: Would you mind adding storedStringsMap as well? Introduced in #31.

**Comment by atn832** [comment]
Looks good, would you mind applying the 2 small changes?

File: test/firebase_storage_mocks_test.dart (hunk 3 above)
Comment (saroad2) [reply test/firebase_storage_mocks_test.dart:179]: Thank you!

File: test/firebase_storage_mocks_test.dart (hunk 2 above)
Comment (saroad2) [reply test/firebase_storage_mocks_test.dart:174]: Thank you!

**Comment by saroad2** [comment]
Hey @atn832 , fixed everything you mentioned in the review. Please let me know if there is something elses needed to be done.

Nevertheless, I would really appriciate it if you could release a new version of *firebase_storage_mock* after you merge this PR🙏 .

File: test/firebase_storage_mocks_test.dart [hunk 6]
```diff
@@ -126,9 +129,58 @@ void main() {
         expect(downloadUrl.contains('some-bucket/o/someimage.png'), isTrue);
       });
     });
+
+    test('Ref listAll', () async {
+      final basePath = 'this/is/basic';
+      final otherPath = 'another/path';
+      final storage = MockFirebaseStorage();
+      await storage.ref(basePath + '/listed_data').putData(randomData(255));
```
Comment (atn832) [inline test/firebase_storage_mocks_test.dart:137]: Much better, thank you!

**Comment by atn832** [comment]
> Nevertheless, I would really appriciate it if you could release a new version of _firebase_storage_mock_ after you merge this PR🙏 .

Sure thing. I just published your feature to https://pub.dev/packages/firebase_storage_mocks/changelog#061. Thanks for the high quality PR.

Btw, in the future, if your work depends urgently on your fixes, you can temporarily change your yaml file to depend on your fork. See https://dart.dev/tools/pub/dependencies#git-packages.

**Pull request merged**